[quant][fx] Support producing reference quantized patterns for dynamic quantization

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89272
* #89271
* #89270
* #89269
* #89268
* __->__ #89248

Summary:
split the is_decomposed logic for `_replace_observer_with_quantize_dequantize_node` in a separate function and added support for dynamic quantization in the decomposed version of this function.

In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

jerryzh168 committed:
[quant][fx] Support producing reference quantized patterns for dynamic quantization

Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -174,7 +195,32 @@ def _replace_observer_with_quantize_dequantize_node(
-    # 2. replace observer node with quant - dequant node
+    # 2. insert choose_qparams op and update the params list for the is_decomposed and
+    # is_dynamic case
+    if is_decomposed and is_dynamic:
```
it's a bit weird that in the previous if block we have

```
elif compute dtype in (...):
  if is_dynamic:
    ...
  
```

and in this one we have

```
if is_decomposed and is_dynamic:
  ...
```

can we keep it consistent?

jerryzh168 replied:
sure

jerryzh168 replied:
it's just a special case for is_decomposed + is_dynamic btw, we don't have other cases, at least for now, but I don't have a strong preference on this

jerryzh168 replied:
actually this feels a bit weird as well when I look at it...:
```
if is_decomposed:
    if is_dynamic:
        ...

# some other code
...
```

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -161,8 +162,28 @@ def _replace_observer_with_quantize_dequantize_node(
+            # quant_min = activation_post_process.quant_min  # type: ignore[attr-defined]
+            # quant_max = activation_post_process.quant_max  # type: ignore[attr-defined]
+            dtype = to_underlying_dtype(dtype)
+            if dtype == torch.uint8:
```
can this be reused with the observer logic in `choose_qparams`?  Ideally there is only one place where this stuff is defined

jerryzh168 replied:
oh this is temporary workaround when the PlacholderObserver doesn't have quant_min/quant_max defined, I think ideally we should just enforce quant_min/quant_max there, choose_qparams takes quant_min/quant_max as argument so we don't have logic to derive quant_min/quant_max from dtype in choose_qparams

jerryzh168 replied:
I can add a follow up PR to make sure PlaceholderObserver has this defined and remove this hack

vkuzo replied:
I think I rate this type of technical debt (hardcoding numbers that are used in mathematical calculations) as "best to never land in the codebase unless it's the final form", it's never temporary once you land it :)  This is riskier than other types of tech debt.

maybe it can be in this PR or in a PR before this one in a stack?

jerryzh168 replied:
yeah sure, I can do that

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -174,7 +195,33 @@ def _replace_observer_with_quantize_dequantize_node(
-    # 2. replace observer node with quant - dequant node
+    # 2. insert choose_qparams op and update the params list for the is_decomposed and
+    # is_dynamic case
+    if is_decomposed:
+        if is_dynamic:
```
now this case matches, but the reader is left with the question of "what about non-decomposed" or "non-dynamic".  Can we add some comments on why we don't have four cases here?

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -165,41 +166,117 @@ def _replace_observer_with_quantize_dequantize_node(
+    # quantization is fixed (based on calibration), we don't need to compute it
+    # dynamically in the graph
+    # with choose_qparams
+    if is_dynamic:
```
Can be done in a future PR, but right now it's not clear what dtypes are supported in what modes without diving into this code. Ideally we would have a helper function or a constant list somewhere since these dtypes are hardcoded already.

andrewor14 replied:
I actually think just setting `is_dynamic = [torch.quint8, torch.qint8, torch.float16]` here is clearer, so the reader doesn't have to trace through the if else cases above.

jerryzh168 replied:
I remember @vkuzo was trying to make this clearer by removing compute_dtype, and just use dtype + is_dynamic flag directly, hopefully things will be clearer after that, @vkuzo can you share the current status on this refactor?

vkuzo replied:
there was still some discussion on ops vs black box backends, I think we just punted on that, that's why I didn't land the PR yet

it's pretty confusing to use `compute_dtype` and `is_dynamic` at the same time in the same function, can we stick to just one of them?  I'd recommend just using `compute_dtype in [torch.quint8, torch.qint8, torch.float16]`, since that's what the current code is using

jerryzh168 replied:
I think right now the condition for is_dynamic is:

```
not (dtype in [torch.quint8, torch.qint8, torch.qint32] and \
            not hasattr(activation_post_process, 'compute_dtype')) and compute_dtype in [torch.quint8, torch.qint8, torch.float16]
```
since "compute_dtype in [torch.quint8, torch.qint8, torch.float16]" in L174 appears in `elif` branch, would it be OK to use this as a condition to replace is_dynamic?

jerryzh168 replied:
I added a TODO to cleanup the condition for first `if`, this is really hard to reason about, maybe it will get better if we have is_dynamic flag

vkuzo replied:
For other PRs in this stack seems ok to fix technical debt later since they are not touching production code, but this PR is directly changing production code, in the part of the codebase that has ranked as really hard to work with for our peers.  I think we should be more careful about adding technical debt here.

Can we either make this better in this PR or just move this code out of the prod path until it's easier to understand?  I think by moving faster on this project by not fixing the debt, we'd be slowing down other developers, so it will be a net negative overall.

jerryzh168 replied:
sounds good, updated

andrewor14 commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -165,41 +166,117 @@ def _replace_observer_with_quantize_dequantize_node(
+    # only needed for is_dynamic because the quantization parameters for static
+    # quantization is fixed (based on calibration), we don't need to compute it
+    # dynamically in the graph
+    # with choose_qparams
```
I feel we can expand on this comment. Can we explain what this code block is doing on a high level? Right now this just explains why we don't need to do this for static

andrewor14 replied:
Actually I've been staring at this code block for a while and I still don't see why we can't just do this in the `if is_decomposed` case in L170. Right now we have a partial `qparams` map there and it's a bit confusing that we expect to populate the scale and zero point later. Is there a reason why this has to be a separate pass?

jerryzh168 replied:
sure basically the scale/zero_point for the dynamic quant case is from choose_qparam node, that we are inserting in this second step, so it's not there in step 1.
high level:
step 1: extract inforrmation from activation_post_process (no graph change)
step 2: (optional) insert choose_qparams, and update the extracted information if needed
step 3: insert q/dq node

will update the comment

andrewor14 commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -132,6 +132,7 @@ def _replace_observer_with_quantize_dequantize_node(
     quantize_op : Optional[Union[Callable, str]] = None
+    is_dynamic = False
     if dtype in [torch.quint8, torch.qint8, torch.qint32] and \
             not hasattr(activation_post_process, 'compute_dtype'):
```
Not related to your changes but can we add some comments to explain what this case is? Is this static quantization?

jerryzh168 replied:
yes, this is static quantization

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -124,16 +124,26 @@ def _replace_observer_with_quantize_dequantize_node(
-    # otherwise, we can convert the observer module call to quantize/dequantize node
+    # otherwise, we can convert the activation_post_process module call to quantize/dequantize node
+    # high level
+    # step 1: extract information for inserting q/dq node from activation_post_process
+    # e.g. quantize_op, quantization parameters (no graph change)
+    # step 2: (optional) insert choose_qparams, and update the extracted information if needed
+    # step 3: insert q/dq node
```
would it help to reorganize the code as 

```
if condition_1:
  # do step 1..n
elif condition_2:
  # do step 1..n
  ...
```

instead of the current organization of

```
# do step 1
if condition_1:
  ...
elif condition_2:
  ...

# do step 2
if condition_1:
  ...
elif condition_2:
  ...
```

vkuzo replied:
that way the reader could just read the specific block they care about and ignore the rest when they are debugging something, versus always being forced to read all the blocks.  One could enforce correctness with tests (already exist) and code reuse as necessary, but I'd bias for simplicity and readability over maximizing reuse.  I think the addition of `is_demposed` is making the current thing too confusing

jerryzh168 replied:
I think we can, there might be some code duplication I think this will be clearer, can I do this after this stack is landed?

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented:
recent update:
split the is_decomposed logic for _replace_observer_with_quantize_dequantize_node in a separate function (_replace_observer_with_quantize_dequantize_node_decomposed) and added support for dynamic quantization in the decomposed version of this function.

also changed the old `_replace_observer_with_quantize_dequantize_node` to also follow the same structure of 1. check dtype and compute_dtype to identify the type of quantization 2. extract qparams + replace observer node with q/dq for that case in the same branch, there are some code duplication but the overall logic is clearer

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
split the is_decomposed logic for `_replace_observer_with_quantize_dequantize_node` in a separate function and added support for dynamic quantization in the decomposed version of this function.

In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support producing reference quantized patterns for dynamic quantization"


Summary:
split the is_decomposed logic for `_replace_observer_with_quantize_dequantize_node` in a separate function and added support for dynamic quantization in the decomposed version of this function.

In case of dynamic quantization, we'll produce the following reference quantized pattern in decomposed mode:
```
x -> choose_qparams -> quantize_per_tensor -> dequantize_per_tensor -> linear
```

Test Plan:
python test/test_quantization.py -k test__convert_to_reference_decomposed_fx_dynamic_quant

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
